stablehlo/reference/Ops.cpp

```diff
@@ -1329,6 +1349,78 @@ Tensor evalRsqrtOp(const Tensor &operand, ShapedType resultType) {
   return result;
 }
 
+SmallVector<Tensor> evalScatterOp(
+    ArrayRef<Tensor> inputs, const Tensor &scatterIndices,
+    ArrayRef<Tensor> updates, const Axes &updateWindowDims,
+    const Axes &insertedWindowDims, const Axes &scatterDimsToOperandDims,
+    Axis indexVectorDim, Region &updateComputation, Scope &scope,
+    ArrayRef<ShapedType> resultTypes) {
+  SmallVector<Tensor> results;
+  for (auto input : inputs) results.push_back(input);
+
+  Axes updateScatterDims;
+  for (auto d : updates[0].getAxes())
+    if (!llvm::is_contained(updateWindowDims, d))
+      updateScatterDims.push_back(d);
+
+  for (auto updateIndexIt = updates[0].index_begin();
+       updateIndexIt != updates[0].index_end(); ++updateIndexIt) {
+    auto updateIndex = *updateIndexIt;
+    Index updateScatterIndex;
+    for (auto d : updateScatterDims)
+      updateScatterIndex.push_back(updateIndex[d]);
+
+    auto startIndicesIndex = updateScatterIndex;
+    if (indexVectorDim < scatterIndices.getRank())
+      startIndicesIndex.insert(startIndicesIndex.begin() + indexVectorDim,
+                               kColon);
+
+    auto startIndex = evalIndex(evalSliceOp(scatterIndices, startIndicesIndex));
+
+    Index fullStartIndex(inputs[0].getRank(), 0);
+    for (auto dOperand : inputs[0].getAxes()) {
+      if (llvm::find(scatterDimsToOperandDims, dOperand) !=
+          scatterDimsToOperandDims.end()) {
+        auto dStart = llvm::find(scatterDimsToOperandDims, dOperand) -
+                      scatterDimsToOperandDims.begin();
+        fullStartIndex[dOperand] = startIndex[dStart];
+      }
+    }
+
+    Index updateWindowIndex;
+    for (auto d : updateWindowDims) updateWindowIndex.push_back(updateIndex[d]);
+
+    Index fullWindowIndex(updateWindowIndex.size() + insertedWindowDims.size(),
+                          0);
+    for (size_t i = 0, oi = 0; i < fullWindowIndex.size(); ++i) {
+      if (llvm::is_contained(insertedWindowDims, i)) continue;
+      fullWindowIndex[i] = updateWindowIndex[oi++];
+    }
+
+    auto resultIndex = fullStartIndex + fullWindowIndex;
+    if (!resultIndex.inBounds(results[0].getShape())) continue;
+
+    SmallVector<Tensor> bodyArgs;
```

Comment:
Should we call this `updateComputationArgs`? Same for other body-related variables.